### PR TITLE
levels!(): enhance performance

### PR DESCRIPTION
**Pull request by alyst**
Enhance the performance of `levels!()` by avoiding redundant checks for duplicate new levels, levels intersection, recoding etc.
These checks become expensive as the number of levels grow.

### CODE REVIEW

**Comment by bkamins** [comment]
@alyst - please allow if the review of the PR might be possibly delayed a bit. If you do not get any feedback in 10 days please bump and I will review (but @nalimilan is a better person to do this and I am sure he eventually will comment on it). Thank you!

File: src/array.jl [hunk 1]
```diff
@@ -769,29 +769,30 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
...
 
-    # prepare a map from new levels to their refs
-    newlv2ref = Dict{eltype(newlevels), Tuple{Int, Int}}()
+    # map each new level to its ref code and number of occurrences
+    newlv2ref = Dict{eltype(newlevels), Pair{Int, Int}}()
     dupnewlvs = similar(newlevels, 0)
     for (i, lv) in enumerate(newlevels)
-        lvref = get(newlv2ref, lv, (i, 0))
-        newlv2ref[lv] = (lvref[1], lvref[2] + 1)
+        lvref = get(newlv2ref, lv, i => 0)
```
Comment (nalimilan) [inline src/array.jl:776]: Why not just `if occursin(newlv2ref, lv); push!(dupnewlvs, lv); continue; end`? To avoid the double dict lookup (when getting and then setting), you could even do `if get!(newlv2ref, lv, i) != i`.

File: src/array.jl [hunk 2]
```diff
@@ -767,41 +767,52 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
...
-            else
-                x > 0 && deleted[x] &&
-                    throw(ArgumentError("cannot remove level $(repr(oldlevels[x])) as it " *
-                                        "is used at position $i. Change the array element " *
-                                        "type to Union{$T, Missing} using convert if you want " *
-                                        "to transform some levels to missing values."))
+            if x > 0 && (oldref2newref[x + 1] == 0)
+                msg = "cannot remove level $(repr(oldlevels[x])) as it is used at position $i"
+                if !(T >: Missing)
+                    msg *= ". Change the array element type to Union{$T, Missing}"
```
Comment (nalimilan) [inline src/array.jl:798]:
```suggestion
                    msg *= ". Change the array element type to Union{$T, Missing}" *
```

File: src/array.jl [hunk 3]
```diff
@@ -767,41 +767,52 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
...
+    for (i, lv) in enumerate(oldlevels)
+        oldref2newref[i + 1] = first(get(newlv2ref, lv, 0 => 0))
+    end
 
     # first pass to check whether, if some levels are removed, changes can be applied without error
-    # TODO: save original levels and undo changes in case of error to skip this step
-    # equivalent to issubset but faster due to JuliaLang/julia#24624
-    if !isempty(setdiff(oldlevels, newlevels))
-        deleted = [!(l in newlevels) for l in oldlevels]
+    if (!(T >: Missing) || !allowmissing) && any(==(0), view(oldref2newref, 2:length(oldref2newref)))
```
Comment (nalimilan) [inline src/array.jl:793]:
```suggestion
    if (!(T >: Missing) || !allowmissing) && any(iszero, @view oldref2newref[2:end])
```

File: src/array.jl [hunk 4]
```diff
@@ -767,41 +767,52 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
...
+    # map each old ref code to new ref code (or 0 if no such level)
+    oldlevels = levels(pool(A))
+    oldref2newref = fill(0, length(oldlevels) + 1)
+    for (i, lv) in enumerate(oldlevels)
+        oldref2newref[i + 1] = first(get(newlv2ref, lv, 0 => 0))
+    end
 
     # first pass to check whether, if some levels are removed, changes can be applied without error
-    # TODO: save original levels and undo changes in case of error to skip this step
-    # equivalent to issubset but faster due to JuliaLang/julia#24624
```
Comment (nalimilan) [inline src/array.jl:779]: The TODO still applies AFAICT. The idea was that we could avoid this first pass and instead do the check in the recoding loop, if we had some code to revert the changes made up to the problematic entry when an error happens.

File: src/array.jl (hunk 1 above)
Comment (alyst) [reply src/array.jl:776]: I wanted to avoid `unique(dupnewlvs)`. But probably your suggestion is less expensive on average and the code is simpler.

File: src/array.jl (hunk 4 above)
Comment (alyst) [reply src/array.jl:779]: I can keep it, but wouldn't always saving the old refs be more memory expensive + saving itself _is_ a pass that affects the cache?

Comment (alyst) [reply src/array.jl:779]: Also, before this PR the checks were more expensive as `oldref2newref` was not available, but now the total cost of the pass is *O(N)* and it only involves integer arrays.

Comment (alyst) [reply src/array.jl:779]: Well, potentially saving the old `A.refs` could be avoid as until the missing level has encountered the changes should be reversible (with `newref2oldref`).
I will commit this variant, you can then decide whether it's worth keeping.

File: src/array.jl [hunk 5]
```diff
@@ -767,41 +767,73 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
...
+            msg = "cannot remove level $(repr(oldlevels[arefs[failedpos]])) as it is used at position $failedpos"
+            if !(T >: Missing)
+                msg *= ". Change the array element type to Union{$T, Missing}" *
+                       " using convert if you want to transform some levels to missing values."
+            elseif !allowmissing
+                msg *= " and allowmissing=false."
+            end
+            throw(ArgumentError(msg))
+        end
+    else # fast pass, either the new missings are allowed or no new missing can occur
```
Comment (nalimilan) [inline src/array.jl:829]:
```suggestion
    else # fast pass, either introducing new missings is allowed or no new missings can occur
```

File: src/array.jl (hunk 4 above)
Comment (nalimilan) [reply src/array.jl:779]: Thanks, that's what I had in mind. Can you add tests to ensure that recoding is correct in the various error cases?

Comment (alyst) [reply src/array.jl:779]: I've added some tests (helped to fix one bug).

**Comment by alyst** [comment]
gentle bump

File: test/12_missingarray.jl [hunk 6]
```diff
@@ -216,13 +216,16 @@ const ≅ = isequal
...
                 @test_throws ArgumentError levels!(x, ["e", "c"])
+                @test x == ["c", "b", "b"]
+                @test levels(x) == ["e", "a", "b", "c"]
+
                 @test levels!(x, ["e", "c"], allowmissing=true) === x
                 @test levels(x) == ["e", "c"]
                 @test x[1] === CategoricalValue(x.pool, 2)
                 @test x[2] === missing
                 @test x[3] === missing
-                @test levels(x) == ["e", "c"]
```
Comment (nalimilan) [inline test/12_missingarray.jl:225]: Also apply these changes to 11_array.jl.

File: test/13_arraycommon.jl [hunk 7]
```diff
@@ -2077,4 +2077,63 @@ end
...
+    @testset "can drop unused levels if element type is $(eltype(x0))" for x0 in (orig, origmissing)
+        x = CategoricalArray(x0)
+        levels!(x, ["E", "A", "B", "C", "D"])
+        @test levels(x) == ["E", "A", "B", "C", "D"]
+        @test x === levels!(x, ["B", "A", "C", "D"])
+        @test x ≅ x0
+        @test levels(x) == ["B", "A", "C", "D"]
+    end
+
+    @testset "CategoricalArray cannot store missings" begin
```
Comment (nalimilan) [inline test/13_arraycommon.jl:2104]:
```suggestion
    @testset "CategoricalArray which cannot store missings" begin
```

File: test/13_arraycommon.jl [hunk 8]
```diff
@@ -2077,4 +2077,63 @@ end
...
+
+        # still throws even if allowmissing=true
+        @test_throws ArgumentError levels!(x, ["B", "A", "C"], allowmissing=true)
+        # check that the x contents have not changed
+        @test x == orig
+        @test pool(x) === oldpool
+        @test levels(x) == ["A", "B", "C", "D"]
+    end
+
+    @testset "CategoricalArray can store missing" begin
```
Comment (nalimilan) [inline test/13_arraycommon.jl:2122]:
```suggestion
    @testset "CategoricalArray which can store missing" begin
```

**Review by nalimilan — COMMENTED** [review]
Sorry for the delay!

File: test/12_missingarray.jl (hunk 6 above)
Comment (alyst) [reply test/12_missingarray.jl:225]: done. I've addressed your latest review (thanks!) and squashed the commits.

File: test/11_array.jl [hunk 9]
```diff
@@ -188,8 +188,16 @@ using CategoricalArrays: DefaultRefType, leveltype
...
+    # check that x is restored correctly when dropping levels is not allowed
+    @test x == ["a", "b", "b"]
+    @test levels(x) == ["b", "a"]
+
     @test_throws ArgumentError levels!(x, ["e", "b"])
+
     @test_throws ArgumentError levels!(x, ["e", "a", "b", "a"])
+    # one again check that x is restored correctly when dropping levels is not allowed
+    @test x == ["a", "b", "b"]
+    @test levels(x) == ["b", "a"]
```
Comment (nalimilan) [inline test/11_array.jl:200]: Sorry to be a pain, but there are equivalent lines in 12_missingarray.jl: can you keep them in sync? :-)

Comment (alyst) [reply test/11_array.jl]: No problem :) Done
